File: csharp/824.goat-latin.cs
/*
 * @lc app=leetcode id=824 lang=csharp
 *
 * [824] Goat Latin
 */

using System.Collections.Generic;
using System.Text;

// @lc code=start
public class Solution {

    private static readonly bool[] Vowels = new bool[] {
        true, false, false, false, true,
        false, false, false, true, false,
        false, false, false, false, true,
        false, false, false, false, false,
        true, false, false, false, false,
        false
    };

    private static IEnumerable<string> GetNext(string S) {
        var sb = new StringBuilder();
        sb.Append("maa");
        foreach (string word in S.Split(' ')) {
            char c = word[0];
            if (Vowels[(c >= 'a' && c <= 'z') ? (int)c - (int)'a' : (int)c -(int)'A']) {
                yield return $"{word}{sb.ToString()}";
            } else {
                yield return $"{word.Substring(1)}{c}{sb.ToString()}";
            }
            
            sb.Append('a');
        }
    }

    public string ToGoatLatin(string S) {
        return string.Join(' ', GetNext(S));
    }
}
// @lc code=end

File: java/824.goat-latin.java
/*
 * @lc app=leetcode id=824 lang=java
 *
 * [824] Goat Latin
 */

// @lc code=start
class Solution {
    public String toGoatLatin(String S) {
        String[] split = S.split(" ");
        StringBuilder sb = new StringBuilder();
        int n = split.length;
        for (int i = 0; i < n; ++i) {
            String word = split[i];
            char c = Character.toLowerCase(word.charAt(0));
            if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u') {
                sb.append(word);
            } else {
                sb.append(word.substring(1));
                sb.append(word.charAt(0));
            }
            sb.append("maa");
            for (int j = 0; j < i; ++j) {
                sb.append('a');
            }
            if (i != n - 1) {
                sb.append(' ');
            }
        }
        return sb.toString();
    }
}
// @lc code=end

